lib/beaker/host/unix/exec.rb

```diff
@@ -49,20 +49,20 @@ def reboot(wait_time=10, max_connection_tries=9, uptime_retries=18)
       sleep(reboot_sleep)
 
       exec(Beaker::Command.new('/bin/systemctl reboot -i || reboot || /sbin/shutdown -r now'), :expect_connection_failure => true)
-    rescue Beaker::Host::RebootFailure => e
+    rescue RuntimeError => e
+      raise Beaker::Host::RebootFailure, "Unexpected exception in reboot: #{e.message}"
```

Comment:
To summarize the previous discussions, I suggest these modifications:
```suggestion
    rescue Beaker::Host::RebootFailure, RuntimeError => e
      raise Beaker::Host::RebootFailure, "Unexpected exception in reboot: #{e.message}",
    rescue Beaker::Host::CommandFailure => e
      raise Beaker::Host::RebootFailure, "Command failed when attempting to reboot: #{e.message}"       
```

Reply:
@op-ct I actually *want* to retry if we trigger one of our exceptions. It turns out that things can fail for all sorts of reasons and somewhere between one and 5 tries later everything will "just work".

Reply:
@trevor-vaughan I could accept that in the abstract, but in this particular instance there's still a problem: it handles nonsensical times in two opposite ways.  I'm comparing:

1. Your reason for using `rescue ArgumentError => e` to avoid the `retry` (given to @ekohl earlier in the review):

   > "This gets thrown **if the regex match on the time doesn't work**. If we get something that's absolute nonsense, we can't compare times."

2.  The reasons for this block's two `Beaker::Host::RebootFailure`s (which don't avoid the `retry`) :

    * The first one is raised *when the regex match for the boot doesn't work*:

      ```ruby
      raise Beaker::Host::RebootFailure, "Could not find system boot time using '#{boot_time_cmd}': '#{original_boot_time_str}'" unless original_boot_time_line
      ```
    * The second one is raised  *when the regex match on the time doesn't work*: 
      ```ruby
      raise Beaker::Host::RebootFailure, "Found no valid times in '#{original_boot_time_line}'" unless original_boot_time_matches
      ```

If "the time regex is nonsense" is the reason to skip the reboot in the first instance, why are we handling it the opposite way in the second instance?

Reply:
@op-ct Oh, that!

Well, in the first instance, I seemed to have a case where this could happen a couple of times but then it would start working (no idea why). In this case, I needed to retry the reboot again because I'm not sure why the command wasn't catching. Note, this happened *extremely* rarely.

I the second instance, the host has just rebooted and literally anything goes! We may have gone back in time, rebooted into the exact same time, something else...? So, in that case, we just loop until things magically work (they usually do the first time but sometimes it takes a handful) or we eventually die and the `raise` throws the original exception up to the user.

Reply:
Thanks;  I think I see what's so weird about this now:

**(TL;DR: None of this does what you think it's doing)**

The are two separate begin/rescue blocks in the `reboot` function:
1. The first block (lines 28–66) records the "original" boot time from the SUT, then reboots it.
2. The second block (lines 69–106) checks the SUT's current boot time to verify that it has rebooted. 

The weirdness: None of what you've just mentioned happens here in this first block, and any exceptions in the second block can't `retry` the reboot:

* The reboot happens on the *final* line (line 51) of the first block. 
  *  That's the only place where the `reboot` function reboots—so if you're already in the second begin/rescue block, you've missed your chance to `retry` it.  
  * More importantly: it's impossible for any exception raised _before_ line 51 to `retry` a reboot, (because it didn't execute the first one).
* So: rescuing either of the `Beaker::Host::RebootFailure` exceptions we've been discussing could never `retry` a reboot, because the reboot happens _after_ they'd be raised.
* They also wouldn't raise in the circumstances you describe:  The purpose of `Beaker::Host::RebootFailure` exceptions (in both blocks) is to retry SUT commands when they don't provide a boot time at all—they don't raise when a boot time was returned, but it's earlier than what you expected.
* The just-rebooted, too-early/same-time problems can't happen here either, because those comparisons necessarily happen _after_ the reboot (in the second block).
* The second block is post-reboot and _does_ compare boot times—but it's impossible to `retry` the reboot from there, too—because the reboot happened in the first begin/rescue block.

So, it looks like the only way this code could actually `retry` to _re_-reboot an SUT is if the `Beaker::Command.new('/bin/systemctl reboot -i ...` at the end of the first block raised an exception internally, which got caught by the open-ended `rescue => e`.  
* Good news: if such an exception was raised, it's very likely to be the `Beaker::Host::CommandFailure` that this patch funnels into `rescue => e`
* Bad news: Even if that happened, it doesn't address your Back to the Future problems, because those are only relevant in the seconds block, and that _can't_ retry the reboot.

If any of this patch helps with your backwards-in-time or same-time scenario, it's a side effect of some kind.  This is why I asked to see an example or a way to make it fail; a lot of this is so unrelated that it would be good to take a step back and validate that the problem you _think_ you have is 1. your actual problem and 2. what you are fixing with these changes.  

Meanwhile, this patch erroneously kneecaps some useful functionality, and—worse yet—allows unrelated exceptions to cause unnecessary re-reboots, potentially several times.  

It would be a mistake to merge this, as it is right now.

Reply:
@trevor-vaughan I'd be happy to look into it.  But I haven't seen a characteristic example of a backwards-in-time failure, so I'm not sure I'd recognize one even if I reproduced it.

I know simp-core runs public acceptance tests in GitLab CI—have there been any failures over there with the problems that this patch is intended to fix?  If so, there should be a log over there we can look at.